lib/src/conflicts.rs

```diff
@@ -473,38 +515,71 @@ pub fn parse_conflict(input: &[u8], num_sides: usize) -> Option<Vec<Merge<BStrin
     }
 }
 
+/// This method handles parsing both JJ-style and Git-style conflict markers,
+/// meaning that switching conflict marker styles won't prevent existing files
+/// with other conflict marker styles from being parsed successfully.
+///
+/// The Git "|||||||" and "=======" sections are treated as "removes" and "adds"
+/// respectively. If they are the first marker present in the conflict, then any
+/// lines before them are treated as an "add" implicitly (because Git doesn't
+/// put any marker before the first "add").
 fn parse_conflict_hunk(input: &[u8]) -> Merge<BString> {
     enum State {
         Diff,
         Minus,
         Plus,
-        Unknown,
+        BeforeMarkers,
     }
-    let mut state = State::Unknown;
+    let mut state = State::BeforeMarkers;
+    // Git-style conflicts have an implicit add section before the first marker
+    let mut before_markers = BString::new(vec![]);
     let mut removes = vec![];
     let mut adds = vec![];
     for line in input.split_inclusive(|b| *b == b'\n') {
         if CONFLICT_MARKER_REGEX.is_match_at(line, 0) {
             match line[0] {
+                // %%%%%%% indicates JJ-style diff section
                 CONFLICT_DIFF_LINE_CHAR => {
                     state = State::Diff;
                     removes.push(BString::new(vec![]));
                     adds.push(BString::new(vec![]));
                     continue;
                 }
+                // ------- indicates JJ-style remove section
                 CONFLICT_MINUS_LINE_CHAR => {
                     state = State::Minus;
                     removes.push(BString::new(vec![]));
                     continue;
                 }
+                // +++++++ indicates JJ-style add section
                 CONFLICT_PLUS_LINE_CHAR => {
                     state = State::Plus;
                     adds.push(BString::new(vec![]));
                     continue;
                 }
+                // ||||||| indicates Git-style remove section
+                CONFLICT_GIT_ANCESTOR_LINE_CHAR => {
+                    // Take lines before the first marker and treat them as an add
+                    if let State::BeforeMarkers = state {
+                        adds.push(std::mem::take(&mut before_markers));
+                    }
+                    state = State::Minus;
+                    removes.push(BString::new(vec![]));
+                    continue;
+                }
+                // ======= indicates Git-style add section
+                CONFLICT_GIT_SEPARATOR_LINE_CHAR => {
+                    // Take lines before the first marker and treat them as an add
+                    if let State::BeforeMarkers = state {
+                        adds.push(std::mem::take(&mut before_markers));
+                    }
```

Comment:
Does this mean that `========` is allowed before `|||||||`? Oh, this is for the "empty side 1 and reordered sections" test? I would consider that invalid instead, but I don't feel strongly.

I wouldn't mind if we disallowed reordered sections in the "snapshot" format either.

Reply:
Yes, it was allowed previously. I updated it to have a stricter parsing logic for Git-style conflicts that requires the 3 sections to be in order. Since the parsing for "snapshot" is the same as "diff", and "diff" currently allows sections to be reordered, I left it as-is for "snapshot".